### PR TITLE
Improve async response and perform vm delete async

### PR DESCRIPTION
**Pull request by pierre-emmanuelJ**
destroy vm reply example

```Golang
vms := allvm // example

resps := asyncTasks(tasks)
for i, r := range resps {
			vm := vms[i]

			if r.error != nil {
				fmt.Fprintf(os.Stderr, "virtual machine %q error: %s\n", vm.Name, r.error.Error()) // nolint: errcheck
				continue
			}

			folder := path.Join(gConfigFolder, "instances", vm.ID.String())

			if _, err := os.Stat(folder); !os.IsNotExist(err) {
				if err := os.RemoveAll(folder); err != nil {
					return err
				}
			}
		}
```

### CODE REVIEW

File: cmd/exo/cmd/request.go [hunk 1]
```diff
@@ -81,13 +83,10 @@ func asyncTasks(tasks []task) ([]interface{}, []error) {
...
 
-	if len(*errors) > 0 {
-		return nil, *errors
-	}
-	return *responses, nil
+	return responses
 }
 
-func execTask(task task, id int, c chan taskStatus, resps *[]interface{}, errors *[]error) {
+func execTask(task task, id int, c chan taskStatus, resps *taskResponse) {
```
Comment (greut) [inline cmd/exo/cmd/request.go]: singular.

File: cmd/exo/cmd/request.go [hunk 2]
```diff
@@ -112,8 +111,19 @@ func execTask(task task, id int, c chan taskStatus, resps *[]interface{}, errors
 
 	if errorReq != nil {
 		c <- taskStatus{id, egoscale.Failure}
-		*errors = append(*errors, errorReq)
+		resps.error = errorReq
+	}
+}
+
+// asyncTaskError return all task with an error
+func asyncTaskError(tasks []taskResponse) []error {
```
Comment (greut) [inline cmd/exo/cmd/request.go]: s/asyncTaskError/filterErrors/ ?

File: cmd/exo/cmd/vm_delete.go [hunk 3]
```diff
@@ -42,15 +43,17 @@ var vmDeleteCmd = &cobra.Command{
...
+		for i, r := range resps {
+			vm := vms[i]
+
+			if r.error != nil {
+				fmt.Fprintf(os.Stderr, "virtual machine %q error: %s\n", vm.Name, r.error.Error()) // nolint: errcheck
+				continue
+			}
 
-			folder := path.Join(gConfigFolder, "instances", resp.ID.String())
+			folder := path.Join(gConfigFolder, "instances", vm.ID.String())
```
Comment (greut) [inline cmd/exo/cmd/vm_delete.go]: :+1:

File: cmd/exo/cmd/vm_delete.go [hunk 4]
```diff
@@ -42,15 +43,17 @@ var vmDeleteCmd = &cobra.Command{
...
+		resps := asyncTasks(tasks)
 
-		for _, r := range resps {
-			resp := r.(*egoscale.VirtualMachine)
+		for i, r := range resps {
+			vm := vms[i]
+
+			if r.error != nil {
+				fmt.Fprintf(os.Stderr, "virtual machine %q error: %s\n", vm.Name, r.error.Error()) // nolint: errcheck
+				continue
```
Comment (greut) [inline cmd/exo/cmd/vm_delete.go]: should you use the `filterErrors` here?

File: cmd/exo/cmd/request.go (hunk 1 above)
Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/request.go]: 👍

File: cmd/exo/cmd/request.go (hunk 2 above)
Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/request.go]: 👍

File: cmd/exo/cmd/request.go [hunk 5]
```diff
@@ -112,8 +111,19 @@ func execTask(task task, id int, c chan taskStatus, resps *[]interface{}, errors
 
 	if errorReq != nil {
 		c <- taskStatus{id, egoscale.Failure}
-		*errors = append(*errors, errorReq)
+		resps.error = errorReq
```
Comment (greut) [inline cmd/exo/cmd/request.go]: `resps.error = fmt.Errorf("failure %s: %s", task.string, errorReq)` ?

Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/request.go]: 👍

**Comment by pierre-emmanuelJ** [comment]
@greut it's look's good for you

File: cmd/exo/cmd/request.go [hunk 6]
```diff
@@ -21,26 +21,31 @@ type taskStatus struct {
...
-	var wg sync.WaitGroup
-	p := mpb.New(mpb.WithWaitGroup(&wg), mpb.WithContext(gContext), mpb.WithWidth(40))
-	wg.Add(len(tasks))
+	var taskWG sync.WaitGroup
+	p := mpb.New(mpb.WithWaitGroup(&taskWG), mpb.WithContext(gContext), mpb.WithWidth(40))
+	taskWG.Add(len(tasks))
+
+	var workerWG sync.WaitGroup
+	workerWG.Add(len(tasks))
+	workerSem := make(chan int, len(tasks))
```
Comment (greut) [inline cmd/exo/cmd/request.go]: if the semaphore is as big as the list of tasks, you're not doing anything different.

File: cmd/exo/cmd/request.go [hunk 7]
```diff
@@ -112,8 +127,21 @@ func execTask(task task, id int, c chan taskStatus, resps *[]interface{}, errors
 
 	if errorReq != nil {
 		c <- taskStatus{id, egoscale.Failure}
-		*errors = append(*errors, errorReq)
+		resp.error = fmt.Errorf("failure %s: %s", task.string, errorReq)
```
Comment (greut) [inline cmd/exo/cmd/request.go]: :+1:

File: cmd/exo/cmd/request.go (hunk 6 above)
Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/request.go]: yes sure so if i want to do all tasks at the same time i can remove semephore !
or for example i can run it 4 by 4 with semaphore for example ?

Comment (greut) [reply cmd/exo/cmd/request.go]: marco says something like 10 or 20.

Comment (greut) [reply cmd/exo/cmd/request.go]: also no magical numbers, so I'd create a proper const for that too.

Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/request.go]: yep yep thx

Comment (pierre-emmanuelJ) [reply cmd/exo/cmd/request.go]: 20 LGTM

**Pull request merged**